ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/TypeHintsAdornmentProvider.fs

```diff
@@ -0,0 +1,125 @@
+                    override x.HasContextMenu = false
+                    override x.ContextMenuTitle = null
+                    override x.ContextMenuItems = null
+                    override x.IsNavigable = false
+                    override x.ExecuteNavigation _ = ()
+                    override x.SelectionRange = Nullable<_>()
+                    override x.IconId = null
+                    override x.IsPreceding = false
+                }
+            | _ -> null
+
+type TypeHighlightingVisitor(fsFile: IFSharpFile, checkResults: FSharpCheckFileResults) =
+    inherit TreeNodeVisitor<IHighlightingConsumer>()
+
+    let tokenNames = ["|>"]
+
+    override x.VisitNode(node, context) =
+        for child in node.Children() do
+            match child with
+            | :? IFSharpTreeNode as treeNode -> treeNode.Accept(x, context)
+            | _ -> ()
+
+    override x.VisitBinaryAppExpr(binding, consumer) =
+        let opExpr = binding.Operator
+        if opExpr.QualifiedName <> "|>" then () else
+
+        match opExpr.Identifier.As<FSharpIdentifierToken>() with
+        | null -> ()
+        | token ->
+
+            let sourceFile = opExpr.GetSourceFile()
+            let coords = sourceFile.Document.GetCoordsByOffset(opExpr.GetTreeEndOffset().Offset)
+            let lineText = sourceFile.Document.GetLineText(coords.Line)
+            use cookie = CompilationContextCookie.GetOrCreate(fsFile.GetPsiModule().GetContextFromModule())
+
+            let getTooltip = checkResults.GetStructuredToolTipText(int coords.Line + 1, int coords.Column, lineText, tokenNames, FSharpTokenTag.Identifier)
+            let (FSharpToolTipText layouts) = getTooltip.RunAsTask()
+
+            // The |> operator shouldn't have any overloads, and it should have two type parameters
+            match layouts with
+            | [ FSharpStructuredToolTipElement.Group [ { TypeMapping = [ _; returnTypeParam ] } ] ] ->
+                // TODO: do something way less hacky here
+                // Trim off the: "'U is " prefix
+                let text = ": " + (showL returnTypeParam).Substring(6)
+
+                let richText = RichText(text, TextStyle(System.Drawing.FontStyle.Underline, Color.Magenta))
+
+                TypeHintHighlighting(richText, binding.RightArgument.GetNavigationRange().EndOffsetRange())
+                |> consumer.AddHighlighting
+            | _ -> ()
+
+        x.VisitNode(binding, consumer)
+
+type TypeHintsHighlightingProcess(fsFile, settings, daemonProcess) =
+    inherit FSharpDaemonStageProcessBase(fsFile, daemonProcess)
+
+    let [<Literal>] opName = "TypeHintsHighlightingProcess"
+
+    override x.Execute(committer) =
+        match fsFile.GetParseAndCheckResults(true, opName) with
+        | None -> ()
+        | Some results ->
+
+        let consumer = FilteringHighlightingConsumer(daemonProcess.SourceFile, fsFile, settings)
+        fsFile.Accept(TypeHighlightingVisitor(fsFile, results.CheckResults), consumer)
```

Comment:
`FSharpDaemonStageProcessBase` inherits the tree visitor too. There's probably no need for a separate types (though, some existing stages/processes may be written this way already).

Reply:
I kept it as a separate type so I could pass in the check results. If I use `TypeHintsHighlightingProcess` as the tree visitor then I would have to do set the check results as a property on this instance and I was trying to avoid mutation where possible.

Any suggestions?

Reply:
There are chances I'd need it to be the same type in future but it's OK to leave it as is for now. Thanks!